client/packages/cli/src/commands/register/vendors/substrate.ts

```diff
@@ -0,0 +1,139 @@
+import fetch from "node-fetch"
+import { Encodings, ApiPromise, WsProvider } from "@t3rn/sdk"
+import { Gateway } from "@/schemas/setup.ts"
+import { spinner } from "../gateway.ts"
+import { colorLogMsg } from "@/utils/log.ts"
+
+export const registerSubstrateVerificationVendor = async (
+  circuit: ApiPromise,
+  gatewayData: Required<Gateway>
+) => {
+  if (!gatewayData.registrationData.parachain) {
+    const target = await ApiPromise.create({
+      provider: new WsProvider(gatewayData.rpc),
+    })
+    return registerRelaychain(circuit, target, gatewayData)
+  }
+
+  return registerParachain(circuit, gatewayData)
+}
+
+const registerRelaychain = async (
+  circuit: ApiPromise,
+  target: ApiPromise,
+  gatewayData: Required<Gateway>
+) => {
+  const portalConsensusData = await fetchPortalConsensusData(
+    circuit,
+    target,
+    gatewayData
+  )
+
+  if (!portalConsensusData) {
+    return
+  }
+
+  const { registrationHeader, authorities, authoritySetId } =
+    portalConsensusData
+
+  spinner.info(
+    colorLogMsg(
+      "INFO",
+      `Registering Block #${registrationHeader.number.toNumber()}`
+    )
+  )
+  spinner.start()
+
+  return circuit
+    .createType("RelaychainRegistrationData", [
+      registrationHeader.toHex(),
+      Array.from(authorities),
+      authoritySetId,
+      gatewayData.registrationData.owner,
+    ])
+    .toHex()
+}
+
+const registerParachain = async (
+  circuit: ApiPromise,
+  gatewayData: Required<Gateway>
+) =>
+  circuit
+    .createType("ParachainRegistrationData", [
+      gatewayData.registrationData.parachain.relayChainId,
+      gatewayData.registrationData.parachain.id,
+    ])
+    .toHex()
+
+const fetchPortalConsensusData = async (
+  circuit: ApiPromise,
+  target: ApiPromise,
+  gatewayData: Required<Gateway>
+) => {
+  const registrationHeight = await fetchLatestAuthoritySetUpdateBlock(
+    gatewayData.subscan
+  )
+
+  if (!registrationHeight) {
+    return
+  }
+
+  const registrationHeader = await target.rpc.chain.getHeader(
+    await target.rpc.chain.getBlockHash(registrationHeight)
+  )
+  const finalityProof = await target.rpc.grandpa.proveFinality(
+    registrationHeight
+  )
+  const authorities =
+    Encodings.Substrate.Decoders.extractAuthoritySetFromFinalityProof(
+      finalityProof
+    )
+  const registratationHeaderHash = await target.rpc.chain.getBlockHash(
+    registrationHeight
+  )
+  const targetAt = await target.at(registratationHeaderHash)
+  const authoritySetId = await targetAt.query.grandpa.currentSetId()
+  return {
+    registrationHeader,
+    authorities: circuit.createType("Vec<AccountId>", authorities),
+    authoritySetId: circuit.createType("SetId", authoritySetId),
+  }
+}
+
+export const fetchLatestAuthoritySetUpdateBlock = async (
+  subscanIndexerApiEndpoint: string
+) => {
+  try {
+    const response = await fetch(
+      subscanIndexerApiEndpoint + "/api/scan/events",
```

Comment:
maybe a comment to docs?
https://support.subscan.io/#events

what does row 20 mean here?